### PR TITLE
Behold, ultimate solution for the union naming problem!

### PR DESCRIPTION
**Pull request by kilohsakul**
uni -> union_nondet and union_product
also moving the definition of intersection to where the other such functions are and renaming the file intersection.cc to product.cc
Don't know yet what to do with the uni in plumbing.

### CODE REVIEW

File: src/nfa/operations.cc [hunk 1]
```diff
@@ -600,7 +600,36 @@ Nfa mata::nfa::minimize(
...
+    return algorithms::product(lhs, rhs, both_final, first_epsilon, prod_map);
+}
+
+Nfa mata::nfa::union_product(const Nfa &lhs, const Nfa &rhs, const Symbol first_epsilon, std::unordered_map<std::pair<State,State>,State> *prod_map) {
+    auto one_final = [&](const State lhs_state,const State rhs_state) {
+        return lhs.final.contains(lhs_state) || rhs.final.contains(rhs_state);
+    };
+
+    if ( lhs.final.empty() || lhs.initial.empty() ) {
+        Nfa result = rhs;
```
Comment (kilohsakul) [inline src/nfa/operations.cc:621]: This does a deep coppy, right?

**Comment by kilohsakul** [comment]
We should probably have a dispatch function for that and use it in tests, right ...
I will do this later.

File: src/nfa/operations.cc (hunk 1 above)
Comment (Adda0) [reply src/nfa/operations.cc:621]: Yes.

**Review by tfiedor — APPROVED** [review]

**Comment by Adda0** [comment]
I rebased and updated this PR onto the current `devel`. I will close this PR and create a new one.

**Pull request closed without merging**